### PR TITLE
8129776: The optimized Stream returned from Files.lines should unmap the mapped byte buffer (if created) when closed

### PR DESCRIPTION
**Pull request by bplb**
Please review this proposed change to unmap the mapped buffer shared between the root and sub-spliterators used in the optimized `Stream` implementation returned by `Files.lines()`. A reference counter is added to track the total number of spliterators sharing the buffer. It is set to 1 when the shared buffer is created, and incremented each time a sub-spliterator is created. It is decremented when traversing begins or when the spliterator is closed. If the counter is zero after it is decremented then the shared buffer is unmapped.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8129776](https://bugs.openjdk.java.net/browse/JDK-8129776): The optimized Stream returned from Files.lines should unmap the mapped byte buffer (if created) when closed 

### Reviewers
 * [Roger Riggs](https://openjdk.java.net/census#rriggs) (@RogerRiggs - **Reviewer**) ⚠️ Review applies to 378214104421e11da5ddbf2cf8aed5c78771f529
 * [Paul Sandoz](https://openjdk.java.net/census#psandoz) (@PaulSandoz - **Reviewer**) ⚠️ Review applies to e2b0219f3b566ada58d2fa35c8dfcab6f0dcb5f3
 * [Alan Bateman](https://openjdk.java.net/census#alanb) (@AlanBateman - **Reviewer**) ⚠️ Review applies to e2b0219f3b566ada58d2fa35c8dfcab6f0dcb5f3

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/2229/head:pull/2229`
`$ git checkout pull/2229`

### CODE REVIEW

File: src/java.base/share/classes/java/nio/file/Files.java [hunk 1]
```diff
@@ -4121,9 +4121,11 @@ public Path next() {
...
             // for a non-zero length file so disallow this case.
             if (length > 0 && length <= Integer.MAX_VALUE) {
-                Spliterator<String> s = new FileChannelLinesSpliterator(fc, cs, 0, (int) length);
-                return StreamSupport.stream(s, false)
-                        .onClose(Files.asUncheckedRunnable(fc));
+                FileChannelLinesSpliterator fcls =
+                    new FileChannelLinesSpliterator(fc, cs, 0, (int) length);
+                return StreamSupport.stream(fcls, false)
+                        .onClose(Files.asUncheckedRunnable(fc))
+                        .onClose(() -> { fcls.close(); });
```
Comment (RogerRiggs) [inline src/java.base/share/classes/java/nio/file/Files.java:4128]: Typically, the brackets {} are omitted for a single expression.

**Review by RogerRiggs — APPROVED** [review]

**Comment by PaulSandoz** [comment]
Looks good. There is some complexity because the `Spliterator` can be obtained from the `Stream` returned by `Files.lines()`. This is not a common case but unfortunately we need to deal with it, otherwise we could just unmap when the `Stream` is closed.  There are edge cases where by splits may occur but they are not traversed, but i don't think we should be concerned about. This is on a good best-effort basis to free up the resource.

Is unmapping via `nioAccess.unmapper(b).unmap();` wired up?

In `Buffer`:
```java
                @Override
                public UnmapperProxy unmapper(ByteBuffer bb) {
                    if (bb instanceof MappedByteBuffer) {
                        return ((MappedByteBuffer)bb).unmapper();
                    } else {
                        return null;
                    }
                }
```

In MappedByteBuffer:
```java
    UnmapperProxy unmapper() {
        return fd != null ?
                new UnmapperProxy() {
                    @Override
                    public long address() {
                        return address;
                    }

                    @Override
                    public FileDescriptor fileDescriptor() {
                        return fd;
                    }

                    @Override
                    public boolean isSync() {
                        return isSync;
                    }

                    @Override
                    public void unmap() {
                        throw new UnsupportedOperationException();
                    }
                } : null;
    }
```
I don't see where `unmapper` is overridden.

**Comment by bplb** [comment]
`unmapper` looks to be overridden in `sun.nio.ch.FileChannelImpl`. Specifically, `map()` obtains the `Unmapper` instance from `mapInternal()`.

**Comment by bplb** [comment]
No, unmapping via `nioAccess.unmapper(b).unmap()` is not wired up. Investigating why.

**Comment by mcimadamore** [comment]
> No, unmapping via `nioAccess.unmapper(b).unmap()` is not wired up. Investigating why.

MappedByteBuffer::unmapper is only used (AFAIK) from AbstractMemorySegmentImpl::ofBuffer. When creating a segment view over a buffer, we need to recover mapping info - but close() on the returned segment is a no-op. So we don't _need_ the unmap() method to be wired up - although it doesn't hurt for it to be wired up, if desired.

**Review by PaulSandoz — APPROVED** [review]

File: src/java.base/share/classes/java/nio/file/FileChannelLinesSpliterator.java [hunk 2]
```diff
@@ -84,19 +88,30 @@
     // Non-null when traversing
     private BufferedReader reader;
 
+    // Number of references to the shared mapped buffer.  Initialized to unity
+    // when the buffer is created by the root spliterator.  Incremented in the
+    // sub-spliterator constructor.  Decremented when 'buffer' transitions from
+    // non-null to null, either when traversing begins or if the spliterator is
+    // closed before traversal.  If the count is zero after decrementing, then
+    // the buffer is unmapped.
+    private AtomicInteger bufRefCount;
```
Comment (AlanBateman) [inline src/java.base/share/classes/java/nio/file/FileChannelLinesSpliterator.java:97]: This should be final but otherwise the changes are okay..

**Pull request closed without merging**